### DISCUSSION TITLE
Support for year-end closing entries for income/expense accounts

### DISCUSSION THREAD

**Discussion by the-solipsist** [ananthakumaran/paisa#80] [Ideas]
## Background
As with traditional double-entry bookkeeping, at the end of the financial year (March 31, in my case), I pass two entries:

1. `FY2016-17 closing balances | Expenses ; closing:2016-2017`
2. `FY2016-17 closing balances | Income ; closing:2016-2017`

These two entries make all expense and income accounts `0`, and transfer the balances to:

1. `Equity:Retained:Expenses:FY16-17` and
2. `Equity:Retained:Income:FY16-17`

## Problem
Because of this, my income and expenses at the end of each financial year are 0, and `paisa` cannot figure out my yearly net income correctly.  Given that shifting balances from income to retained-income is a common practice in double-entry bookkeeping, I think it should be supported.

## Potential solution
It'd be useful if there was some way of _excluding_ specific entries (via tags, e.g.) when calculating net income.  That was I can exclude all entries with the tag `closing`.

**Comment by ananthakumaran** [comment]
How do you organize files? If you keep these entries in separate files, then it might be possible to have two main ledger files, one you normally use with cli and one via paisa.

paisa-main.ledger
```
include 2021.ledger
include 2022.ledger
```

main.ledger
```
include paisa-main.ledger
include accountcloseentries.ledger
```

Coming back to the issue, using tags to exclude entries is an interesting idea, I will think about it.

**Comment by the-solipsist** [comment]
That's a very interesting suggestion. It'll require me to:
a) separate those entries;
b) create a different default journal for hledger, which includes those entries;
c) create a different journal for paisa which excludes those entries.
```
$ find ./*.journal  -printf "%f\n"
accounts.journal  ; list of all accounts
commodities.journal  ; list of all commodities
foreignholdings.journal  ; foreign currency in cash
fy2017-18.journal  ; first line: include meta.journal
fy2018-19.journal  ; first line: include fy2017-18.journal
fy2019-20.journal
fy2020-21.journal
fy2021-22.journal
fy2022-23.journal
fy2023-24.journal
latest.journal  ; symlink to latest journal (currently, fy2023-24.journal)
meta.journal  ; includes {rates,accounts.commodities}.journal
meta-print.journal  ; same as meta, but without commas in numbers.
rates.journal ; includes all rates_*.journals except rates_mf_all.journal
rates_crypto.journal
rates_currency.journal
rates_equivalents.journal
rates_gold.journal
rates_mf.journal
rates_mf_all.journal  ;  includes daily mutual fund prices, generally only needed for plotting.
rates_nps.journal
rates_rewards.journal
rates_stocks.journal
rates_tangibles.journal
rewards.journal  ; rewards / loyalty points / frequent flyer miles, etc., tracked separately.
```
The first line of each fyxxxx-xx.journal is an include statement with the filename of the previous year's journal.  Thus all years' journals are linked.

`LEDGER_FILE` is set to `~/accounts/latest.journal`

**Comment by adamantike** [comment]
Following the setup provided by https://github.com/adept/full-fledged-hledger triggers this behavior, where charts in Paisa go down to zero on December 31st, and then back the the correct value on January 1st.

Here's the documentation on how the project generates those opening/closing balance entries: https://github.com/adept/full-fledged-hledger/wiki/Getting-full-history-of-the-account#on-the-opening-balances

> How do you organize files? If you keep these entries in separate files, then it might be possible to have two main ledger files, one you normally use with cli and one via paisa.

The `all.journal` file includes all `YYYY.journal` and `YYYY-closing.journal` files. However, each `YYYY.journal` file has an include for `YYYY-opening.journal` to be self-sufficient and be able to query a single year if needed. Trying to have a separate file specifically for Paise would be challenging because of that, I imagine.

**Comment by thursenfuerst** [comment]
For some necromancy of this idea, I also follow the full-fledged-hledger workflow and would love to see some kind of compatibility workaround.
Tag exclusion seems like a good compromise for that I would assume
